puppet/zulip/manifests/sharding.pp

```diff
@@ -0,0 +1,45 @@
+class zulip::sharding {
+  include zulip::base
+  include zulip::common
+  include zulip::nginx
+
+  # These two execs below are rather ugly logic to have in puppet,
+  # but the advantage is that if the format of the config file generated
+  # by the sharding script changes, they will automatically get re-generated
+  # by puppet apply in the process of upgrading zulip, without requiring
+  # the administrator to think about this step.
+  exec { 'sharding_script':
+    subscribe => File['/etc/zulip/zulip.conf'],
+    command   => '/home/zulip/deployments/current/scripts/lib/sharding.py',
+    onlyif    => 'test -f /home/zulip/deployments/current/scripts/lib/sharding.py\
+  -a ! -f /home/zulip/deployments/next/scripts/lib/sharding.py',
+  }
+  exec { 'sharding_script_next':
+    subscribe => File['/etc/zulip/zulip.conf'],
+    command   => '/home/zulip/deployments/next/scripts/lib/sharding.py',
+    onlyif    => 'test -f /home/zulip/deployments/next/scripts/lib/sharding.py',
```

Comment:
What's the background for the `test -f` lines here?

Reply:
We as a policy matter generally don't access `/home/zulip/deployments` from Puppet manifests, because it's dangerous if you're running puppet from a given checkout and we end up reaching into a different one.

I think there's two good solutions:
* Pass into `puppet` somehow which checkpoint `zulip-puppet-apply` was run from; I haven't researched but surely there's a good way to do it.  Might be cleanest.
* Moving `sharding.py` to live under `puppet/` and having it be accessed that way, like the Nagios scripts.

Reply:
Yeah, apparently we already provide puppet with `zulip_scripts_path` in zulip-puppet-apply, so I was able to use that and get rid of this whole ugly block.